File: Leetcode/Java/20. Valid Parentheses/solution.java
// https://leetcode.com/problems/valid-parentheses/description/

class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> parenthesisHash = new HashMap<>();
        parenthesisHash.put('}', '{');
        parenthesisHash.put(']', '[');
        parenthesisHash.put(')', '(');

        Deque<Character> stack = new ArrayDeque<>();

        for (int idx = 0; idx < s.length(); idx++) {
            char p = s.charAt(idx);

            if (parenthesisHash.containsKey(p)) {
                if (stack.isEmpty() || stack.pop() != parenthesisHash.get(p)) {
                    return false;
                }
                continue;
            }

            stack.add(p);
        }

        // Check remain open parenthesis
        return stack.isEmpty();
    }
}

File: Leetcode/Typescript/20. Valid Parentheses/solution.ts
// https://leetcode.com/problems/valid-parentheses/description/

function isValid(s: string): boolean {
  const parenthesis = { "}": "{", "]": "[", ")": "(" },
    stack: string[] = [];

  for (const p of s) {
    if (!parenthesis[p]) {
      stack.push(p);
    } else if (stack.length === 0 || stack.pop() !== parenthesis[p]) {
      return false;
    }
  }

  return stack.length === 0 ? true : false;
}
